### ISSUE TITLE
cloud-init misconfigure the network on SLES

### ISSUE THREAD

**Issue by ubuntu-server-builder** [canonical/cloud-init#3448]
This bug was originally filed in Launchpad as [LP: #1843634](https://bugs.launchpad.net/cloud-init/+bug/1843634)
<details>
<summary>Launchpad details</summary>
<pre>
affected_projects = ['cloud-init (Suse)']
assignee = rjschwei
assignee_name = Robert Schweikert
date_closed = 2019-12-19T23:00:30.263101+00:00
date_created = 2019-09-11T18:40:06.185838+00:00
date_fix_committed = 2019-09-26T19:52:08.211281+00:00
date_fix_released = 2019-12-19T23:00:30.263101+00:00
id = 1843634
importance = high
is_complete = True
lp_url = https://bugs.launchpad.net/cloud-init/+bug/1843634
milestone = None
owner = momousta
owner_name = Moustafa Moustafa
private = False
status = fix_released
submitter = momousta
submitter_name = Moustafa Moustafa
tags = []
duplicates = []
</pre>
</details>

_Launchpad user **Moustafa Moustafa(momousta)** wrote on 2019-09-11T18:40:06.185838+00:00_

cloud-init's sysconfig renderer injects the cloud-init header even
if no DNS configuration is provided in network-config.

cloud-init.service may need to update service file with additional
wicked service names to ensure that networking is up before starting.

[Original Description]
I reproduced the issue on an Azure VM with SLES12 SP4 and cloud-init 19.1.
The DNS is unreachable when cloud-init takes the responsibility of configuring the network. No nameservers or search domains are added to the /etc/resolv.conf as following:
; Created by cloud-init on instance boot automatically, do not edit.
;
### /etc/resolv.conf file autogenerated by netconfig!
#
# Before you change this file manually, consider to define the
# static DNS configuration using the following variables in the
# /etc/sysconfig/network/config file:
#     NETCONFIG_DNS_STATIC_SEARCHLIST
#     NETCONFIG_DNS_STATIC_SERVERS
#     NETCONFIG_DNS_FORWARDER
# or disable DNS configuration updates via netconfig by setting:
#     NETCONFIG_DNS_POLICY=''
#
# See also the netconfig(8) manual page and other documentation.
#
# Note: Manual change of this file disables netconfig too, but
# may get lost when this file contains comments or empty lines
# only, the netconfig settings are same with settings in this
# file and in case of a "netconfig update -f" call.
#
### Please remove (at least) this line when you modify the file!

I also attached the "/etc/sysconfig/network/config" in the first comment for your reference:

When I disable the network configuration in cloud-init and leave it for netconfig, the /etc/resolv.conf is correctly populated with the search domain and the nameserver and the DNS is reachable. Here's the contents of the /etc/resolv.conf:
### /etc/resolv.conf file autogenerated by netconfig!
#
# Before you change this file manually, consider to define the
# static DNS configuration using the following variables in the
# /etc/sysconfig/network/config file:
#     NETCONFIG_DNS_STATIC_SEARCHLIST
#     NETCONFIG_DNS_STATIC_SERVERS
#     NETCONFIG_DNS_FORWARDER
# or disable DNS configuration updates via netconfig by setting:
#     NETCONFIG_DNS_POLICY=''
#
# See also the netconfig(8) manual page and other documentation.
#
# Note: Manual change of this file disables netconfig too, but
# may get lost when this file contains comments or empty lines
# only, the netconfig settings are same with settings in this
# file and in case of a "netconfig update -f" call.
#
### Please remove (at least) this line when you modify the file!
search xkf00b0rtzgejkug4xc2pcinre.xx.internal.cloudapp.net
nameserver 168.63.129.16

When I tried to populate the network config dictionary that's built by DataSourceAzure with a default nameserver "168.63.129.16" and search domain "xkf00b0rtzgejkug4xc2pcinre.xx.internal.cloudapp.net", The DNS was reachable. But It's my understanding that cloud-init should be able to figure out this nameserver and the search domain the same way netconfig does.

Another issue is the eth0 interface is not brought up automatically even though the contents of the file "/etc/sysconfig/network/ifcfg-eth0" seems correct
# Created by cloud-init on instance boot automatically, do not edit.
#
BOOTPROTO=dhcp
DEVICE=eth0
HWADDR=00:0d:3a:06:1e:04
NM_CONTROLLED=no
ONBOOT=yes
STARTMODE=auto
TYPE=Ethernet
USERCTL=no

I had to bring it up by executing "ifup eth0" from Azure.py code after the network config is applied. This way I was able to ssh into the VM.

Here's also the contents of "etc/udev/rules.d/85-persistent-net-cloud-init.rules":SUBSYSTEM=="net", ACTION=="add", DRIVERS=="?*", ATTR{address}=="00:0d:3a:6d:e4:53", NAME="eth0"

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Moustafa Moustafa(momousta)** wrote on 2019-09-11T18:40:06.185838+00:00_


_Launchpad attachments_: [cloud-init.tar.gz](https://api.launchpad.net/devel/cloud-init/+bug/1843634/+attachment/5288148/data)

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Moustafa Moustafa(momousta)** wrote on 2019-09-11T19:02:37.726597+00:00_


_Launchpad attachments_: [config](https://api.launchpad.net/devel/cloud-init/+bug/1843634/+attachment/5288152/data)

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Ryan Harper(raharper)** wrote on 2019-09-12T01:11:41.304297+00:00_

OK, thanks for the logs.  Could you re-attach those running via sudo (or as root)?  The default user on SLES does not have permissions to read the journal.

What I see so far looks like networking did not come up after cloud-init-local.service completes and writes out a network config.

2019-09-11 18:00:15,242 - stages.py[INFO]: Applying network configuration from ds bringup=False: {'ethernets': {'eth0': {'set-name': 'eth0', 'match': {'macaddress': u'00:0d:3a:6e:6f:8f'}, 'dhcp4': True}}, 'version': 2}

This results in the following files being written:

% cat test_azure_sles/etc/sysconfig/network/ifcfg-eth0 
# Created by cloud-init on instance boot automatically, do not edit.
#
BOOTPROTO=dhcp
DEVICE=eth0
HWADDR=00:0d:3a:6e:6f:8f
NM_CONTROLLED=no
ONBOOT=yes
STARTMODE=auto
TYPE=Ethernet
USERCTL=no

Upstream cloud-init on SLES does not generate/update /etc/resolv.conf  but in the logs the cloud-init in does:

2019-09-11 18:00:15,246 - util.py[DEBUG]: Writing to /etc/sysconfig/network/ifcfg-eth0 - wb: [644] 191 bytes
2019-09-11 18:00:15,247 - util.py[DEBUG]: Reading from /etc/resolv.conf (quiet=False)
2019-09-11 18:00:15,247 - util.py[DEBUG]: Read 795 bytes from /etc/resolv.conf
2019-09-11 18:00:15,247 - util.py[DEBUG]: Writing to /etc/resolv.conf - wb: [644] 866 bytes

At first, I thought maybe it was missing this commit:

% git show b74ebca563a21332b29482c8029e7908f60225a4
commit b74ebca563a21332b29482c8029e7908f60225a4
Author: Robert Schweikert <rjschwei@suse.com>
Date:   Wed Jan 23 22:35:32 2019 +0000

    net/sysconfig: do not write a resolv.conf file with only the header.
    
    Writing the file with no dns information may prevent distro tools
    from writing a resolv.conf file with dns information obtained from
    a dhcp server.

diff --git a/cloudinit/net/sysconfig.py b/cloudinit/net/sysconfig.py
index ae41f7b..fd8e501 100644
--- a/cloudinit/net/sysconfig.py
+++ b/cloudinit/net/sysconfig.py
@@ -557,6 +557,8 @@ class Renderer(renderer.Renderer):
             content.add_nameserver(nameserver)
         for searchdomain in network_state.dns_searchdomains:
             content.add_search_domain(searchdomain)
+        if not str(content):
+            return None
         header = _make_header(';')
         content_str = str(content)
         if not content_str.startswith(header):
@@ -666,7 +668,8 @@ class Renderer(renderer.Renderer):
             dns_path = util.target_path(target, self.dns_path)
             resolv_content = self._render_dns(network_state,
                                               existing_dns_path=dns_path)
-            util.write_file(dns_path, resolv_content, file_mode)
+            if resolv_content:
+                util.write_file(dns_path, resolv_content, file_mode)
         if self.networkmanager_conf_path:
             nm_conf_path = util.target_path(target,
                                             self.networkmanager_conf_path)
diff --git a/tests/unittests/test_net.py b/tests/unittests/test_net.py
index d679e92..5313d2d 100644
--- a/tests/unittests/test_net.py
+++ b/tests/unittests/test_net.py
@@ -2098,6 +2098,10 @@ TYPE=Ethernet
 USERCTL=no
 """
         self.assertEqual(expected, found[nspath + 'ifcfg-interface0'])
+        # The configuration has no nameserver information make sure we
+        # do not write the resolv.conf file
+        respath = '/etc/resolv.conf'
+        self.assertNotIn(respath, found.keys())
 
     def test_config_with_explicit_loopback(self):
         ns = network_state.parse_net_config_data(CONFIG_V1_EXPLICIT_LOOPBACK)
@@ -2456,6 +2460,10 @@ TYPE=Ethernet
 USERCTL=no
 """
         self.assertEqual(expected, found[nspath + 'ifcfg-interface0'])
+        # The configuration has no nameserver information make sure we
+        # do not write the resolv.conf file
+        respath = '/etc/resolv.conf'
+        self.assertNotIn(respath, found.keys())
 
     def test_config_with_explicit_loopback(self):
         ns = network_state.parse_net_config_data(CONFIG_V1_EXPLICIT_LOOPBACK)

But, I believe that is in 19.1 (or likely patched in the distro version).
Later in the boot, we can see that networking didn't actually come up as Azure datasource can't find a lease file and then goes into some sort of fallback mode which tries to bring up networking (it does) but not with dhcp which is why you're missing DNS (it's provided via option to the DHCP response.


2019-09-11 18:00:15,946 - azure.py[DEBUG]: Unable to find endpoint in dhclient logs.  Falling back to check lease files
2019-09-11 18:00:15,946 - azure.py[DEBUG]: Looking for endpoint in lease file /var/lib/dhcp/dhclient.eth0.leases
2019-09-11 18:00:15,946 - handlers.py[DEBUG]: start: azure-ds/_get_value_from_leases_file: _get_value_from_leases_file
2019-09-11 18:00:15,946 - util.py[DEBUG]: Reading from /var/lib/dhcp/dhclient.eth0.leases (quiet=False)
2019-09-11 18:00:15,947 - azure.py[ERROR]: Failed to read /var/lib/dhcp/dhclient.eth0.leases: [Errno 2] No such file or directory: '/var/lib/dhcp/dhclient.eth0.leases'
2019-09-11 18:00:15,959 - handlers.py[DEBUG]: finish: azure-ds/_get_value_from_leases_file: SUCCESS: _get_value_from_leases_file
2019-09-11 18:00:15,959 - util.py[DEBUG]: Running command ['ifconfig'] with allowed return codes [0] (shell=False, capture=True)
2019-09-11 18:00:16,020 - azure.py[DEBUG]: ifconfig out: lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:2 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:140 (140.0 b)  TX bytes:140 (140.0 b)

, err: 
2019-09-11 18:00:16,020 - util.py[DEBUG]: Running command ['route', '-n'] with allowed return codes [0] (shell=False, capture=True)
2019-09-11 18:00:16,093 - azure.py[DEBUG]: route out: Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
, err: 
2019-09-11 18:00:16,093 - util.py[DEBUG]: Running command ['ip', 'a'] with allowed return codes [0] (shell=False, capture=True)
2019-09-11 18:00:16,095 - azure.py[DEBUG]: ip out: 1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc mq state DOWN group default qlen 1000
    link/ether 00:0d:3a:6e:6f:8f brd ff:ff:ff:ff:ff:ff
, err: 
2019-09-11 18:00:16,095 - util.py[DEBUG]: Running command ['ifup', 'eth0'] with allowed return codes [0] (shell=False, capture=True)
2019-09-11 18:00:31,824 - azure.py[DEBUG]: ifup out: eth0            up
, err: 
2019-09-11 18:00:31,824 - util.py[DEBUG]: Running command ['ip', '-o', 'route', 'list'] with allowed return codes [0] (shell=False, capture=True)
2019-09-11 18:00:31,827 - azure.py[DEBUG]: ip out: default via 10.0.0.1 dev eth0 proto dhcp 
10.0.0.0/24 dev eth0 proto kernel scope link src 10.0.0.17 
168.63.129.16 via 10.0.0.1 dev eth0 proto dhcp 
169.254.169.254 via 10.0.0.1 dev eth0 proto dhcp 
, err: 
2019-09-11 18:00:31,828 - util.py[DEBUG]: Running command ['ifconfig'] with allowed return codes [0] (shell=False, capture=True)
2019-09-11 18:00:31,830 - azure.py[DEBUG]: ifconfig out: eth0      Link encap:Ethernet  HWaddr 00:0D:3A:6E:6F:8F  
          inet addr:10.0.0.17  Bcast:10.0.0.255  Mask:255.255.255.0
          inet6 addr: fe80::20d:3aff:fe6e:6f8f/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:8 errors:0 dropped:0 overruns:0 frame:0
          TX packets:29 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:2709 (2.6 Kb)  TX bytes:3373 (3.2 Kb)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:6 errors:0 dropped:0 overruns:0 frame:0
          TX packets:6 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:468 (468.0 b)  TX bytes:468 (468.0 b)

, err: 
2019-09-11 18:00:31,831 - util.py[DEBUG]: Running command ['route', '-n'] with allowed return codes [0] (shell=False, capture=True)
2019-09-11 18:00:31,834 - azure.py[DEBUG]: route out: Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.0.0.1        0.0.0.0         UG    0      0        0 eth0
10.0.0.0        0.0.0.0         255.255.255.0   U     0      0        0 eth0
168.63.129.16   10.0.0.1        255.255.255.255 UGH   0      0        0 eth0
169.254.169.254 10.0.0.1        255.255.255.255 UGH   0      0        0 eth0
, err: 
2019-09-11 18:00:31,834 - util.py[DEBUG]: Running command ['ip', 'a'] with allowed return codes [0] (shell=False, capture=True)
2019-09-11 18:00:31,837 - azure.py[DEBUG]: ip out: 1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0d:3a:6e:6f:8f brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.17/24 brd 10.0.0.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::20d:3aff:fe6e:6f8f/64 scope link 
       valid_lft forever preferred_lft forever
, err: 
2019-09-11 18:00:31,837 - azure.py[WARNING]: No lease found; using default endpoint
2019-09-11 18:00:31,837 - azure.py[DEBUG]: Azure endpoint found at 168.63.129.16
2019-09-11 18:00:31,837 - handlers.py[DEBUG]: finish: azure-ds/find_endpoint: SUCCESS: find_endpoint



So, I'd like to also the contents of:

/etc/resolv.conf
/etc/sysconfig/network/ifcfg-eth0

And running sudo cloud-init collect-logs so we can get the journal, which should reveal why SLES's networking service didn't come online before cloud-init.service started.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Robert Schweikert(rjschwei)** wrote on 2019-09-12T11:37:18.382074+00:00_

The original description certainly indicates that cloud-init still writes an "empty" resolv.conf file. However this issues has been addressed for a while thus something does not fit together. It makes no sense that this would only show up in Azure and in the bug that triggered investigation and ultimately led to a fix it has been confirmed as fixed.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Ryan Harper(raharper)** wrote on 2019-09-12T18:01:39.515396+00:00_

Thank you for the logs.

> Who is responsible for adding the name server and the search domain to the network_state ?

For Azure, there will not be any nameserver/search values added to network state as Azure's network-config is to DHCP only on eth0 and The DHCP _client_ from the OS runs will update /etc/resolv.conf

Looking at the log, it appears that Wickedd does run DHCP, but cloud-init.service started too early.


# here wicked starts
Sep 11 17:59:55.140087 localhost systemd[1]: Started System Logging Service.
Sep 11 17:59:55.796842 localhost systemd[1]: Started wicked DHCPv4 supplicant service.
Sep 11 17:59:55.805347 localhost systemd[1]: Started wicked DHCPv6 supplicant service.
Sep 11 17:59:55.812068 localhost systemd[1]: Started wicked AutoIPv4 supplicant service.
Sep 11 17:59:55.819849 localhost systemd[1]: Starting wicked network management service daemon...
Sep 11 17:59:55.933440 localhost systemd[1]: Started wicked network management service daemon.
Sep 11 17:59:55.940875 localhost systemd[1]: Starting wicked network nanny service...
Sep 11 17:59:56.022347 localhost systemd[1]: Started wicked network nanny service.

# then cloud-init-local.service
Sep 11 18:00:07.462643 localhost cloud-init[769]: Cloud-init v. 19.1 running 'init-local' at Wed, 11 Sep 2019 18:00:07 +0000. Up 37.63 seconds.

# then cloud-init-local runs dhclient to bring the network up to crawl Azure IMDS and write out
# network config file (/etc/sysconfig/network/ifcfg-eth0
Sep 11 18:00:09.304202 localhost dhclient[834]: Internet Systems Consortium DHCP Client 4.3.3
Sep 11 18:00:09.304631 localhost dhclient[834]: Copyright 2004-2015 Internet Systems Consortium.
Sep 11 18:00:09.304827 localhost dhclient[834]: All rights reserved.
Sep 11 18:00:09.305018 localhost dhclient[834]: For info, please visit https://www.isc.org/software/dhcp/
Sep 11 18:00:09.305262 localhost dhclient[834]: 
Sep 11 18:00:09.415443 localhost dhclient[834]: Listening on LPF/eth0/00:0d:3a:6e:6f:8f
Sep 11 18:00:09.415458 localhost dhclient[834]: Sending on   LPF/eth0/00:0d:3a:6e:6f:8f
Sep 11 18:00:09.415938 localhost kernel: NET: Registered protocol family 17
Sep 11 18:00:09.415496 localhost dhclient[834]: Sending on   Socket/fallback
Sep 11 18:00:09.415533 localhost dhclient[834]: DHCPDISCOVER on eth0 to 255.255.255.255 port 67 interval 4 (xid=0x1d3935b9)
Sep 11 18:00:09.431456 localhost dhclient[834]: DHCPOFFER from 168.63.129.16
Sep 11 18:00:14.432053 localhost dhclient[834]: DHCPREQUEST on eth0 to 255.255.255.255 port 67 (xid=0x1d3935b9)
Sep 11 18:00:14.449380 localhost dhclient[834]: DHCPACK from 168.63.129.16 (xid=0x1d3935b9)
Sep 11 18:00:14.460401 localhost dhclient[834]: bound to 10.0.0.17 -- renewal in 4294967295 seconds.

# Now cloud-init.service runs and *expects* networking to be up, but it isn't
Sep 11 18:00:15.621036 momousta-vm-945207600 cloud-init[861]: Cloud-init v. 19.1 running 'init' at Wed, 11 Sep 2019 18:00:15 +0000. Up 45.81 seconds.
Sep 11 18:00:15.636134 momousta-vm-945207600 cloud-init[861]: ci-info: +++++++++++++++++++++++++++Net device info++++++++++++++++++++++++++++
Sep 11 18:00:15.648761 momousta-vm-945207600 cloud-init[861]: ci-info: +--------+-------+-----------+-----------+-------+-------------------+
Sep 11 18:00:15.668090 momousta-vm-945207600 cloud-init[861]: ci-info: | Device |   Up  |  Address  |    Mask   | Scope |     Hw-Address    |
Sep 11 18:00:15.680368 momousta-vm-945207600 cloud-init[861]: ci-info: +--------+-------+-----------+-----------+-------+-------------------+
Sep 11 18:00:15.704126 momousta-vm-945207600 cloud-init[861]: ci-info: |  eth0  | False |     .     |     .     |   .   | 00:0d:3a:6e:6f:8f |
Sep 11 18:00:15.720137 momousta-vm-945207600 cloud-init[861]: ci-info: |   lo   |  True | 127.0.0.1 | 255.0.0.0 |  host |         .         |
Sep 11 18:00:15.734575 momousta-vm-945207600 cloud-init[861]: ci-info: |   lo   |  True |  ::1/128  |     .     |  host |         .         |
Sep 11 18:00:15.745009 momousta-vm-945207600 cloud-init[861]: ci-info: +--------+-------+-----------+-----------+-------+-------------------+
Sep 11 18:00:15.754527 momousta-vm-945207600 cloud-init[861]: ci-info: +++++++++++++++++++Route IPv6 info+++++++++++++++++++
Sep 11 18:00:15.765326 momousta-vm-945207600 cloud-init[861]: ci-info: +-------+-------------+---------+-----------+-------+
Sep 11 18:00:15.772241 momousta-vm-945207600 cloud-init[861]: ci-info: | Route | Destination | Gateway | Interface | Flags |
Sep 11 18:00:15.780449 momousta-vm-945207600 cloud-init[861]: ci-info: +-------+-------------+---------+-----------+-------+
Sep 11 18:00:15.787789 momousta-vm-945207600 cloud-init[861]: ci-info: +-------+-------------+---------+-----------+-------+
Sep 11 18:00:15.962107 momousta-vm-945207600 cloud-init[861]: 2019-09-11 18:00:15,947 - azure.py[ERROR]: Failed to read /var/lib/dhcp/dhclient.eth0.leases: [Errno 2] No such file or directory: '/var/lib/dhcp/dhclient.eth0.leases'

# and now we see that wickedd completes DHCP after cloud-init.service ran
ep 11 18:00:16.651842 momousta-vm-945207600 wickedd[811]: ni_process_reap: process 886 has not exited yet; now doing a blocking waitpid()
Sep 11 18:00:16.706739 momousta-vm-945207600 wickedd-dhcp4[770]: eth0: Request to acquire DHCPv4 lease with UUID 3036795d-66f7-0900-2b03-000002000000
Sep 11 18:00:16.707689 momousta-vm-945207600 wickedd-dhcp6[766]: eth0: Request to acquire DHCPv6 lease with UUID 3036795d-66f7-0900-2b03-000003000000 in mode auto
Sep 11 18:00:17.813712 momousta-vm-945207600 wickedd-dhcp4[770]: eth0: Committed DHCPv4 lease with address 10.0.0.17 (lease time 4294967295 sec, renew in 2147483647 sec, rebind in 536870911 sec)
Sep 11 18:00:18.759766 momousta-vm-945207600 dns-resolver[953]: ATTENTION: You have modified /etc/resolv.conf. Leaving it untouched...
Sep 11 18:00:18.819333 momousta-vm-945207600 dns-resolver[955]: You can find my version in /etc/resolv.conf.netconfig


So two things here:

1) cloud-init.service already has an After=wicked.service ... we need to confirm that is the correct service name that cloud-init should wait for

2) wicked doesn't like the modifications to resolv.conf that cloud-init makes
Sep 11 18:00:18.759766 momousta-vm-945207600 dns-resolver[953]: ATTENTION: You have modified /etc/resolv.conf. Leaving it untouched...
Sep 11 18:00:18.819333 momousta-vm-945207600 dns-resolver[955]: You can find my version in /etc/resolv.conf.netconfig


For (2) it appears the commit we have b74ebca563a21332b29482c8029e7908f60225a4 doesn't take into account a existing and non-empty template resolv.conf, as we load the existing content which then means we always write out the file (even with no nameservers or search paths to write).

def _render_dns(network_state, existing_dns_path=None):                    
        content = resolv_conf.ResolvConf("")                                   
        if existing_dns_path and os.path.isfile(existing_dns_path):            
            content = resolv_conf.ResolvConf(util.load_file(existing_dns_path))
        for nameserver in network_state.dns_nameservers:                       
            content.add_nameserver(nameserver)                                 
        for searchdomain in network_state.dns_searchdomains:                   
            content.add_search_domain(searchdomain)                            
        if not str(content):                                                   
            return None                                                        
        header = _make_header(';')                                             
        content_str = str(content)                                             
        if not content_str.startswith(header):                                 
            content_str = header + '\n' + content_str                          
        return content_str       

I think we need to instead skip writing resolv.conf unless we have content to add

% git diff cloudinit/net/sysconfig.py
diff --git a/cloudinit/net/sysconfig.py b/cloudinit/net/sysconfig.py
index be5dede..12c6e36 100644
--- a/cloudinit/net/sysconfig.py
+++ b/cloudinit/net/sysconfig.py
@@ -585,7 +585,11 @@ class Renderer(renderer.Renderer):
             content.add_nameserver(nameserver)
         for searchdomain in network_state.dns_searchdomains:
             content.add_search_domain(searchdomain)
-        if not str(content):
+        # content may be non-empty if an existing were present so also
+        # skip writing a resolv conf if we don't have any nameservers or
+        # search paths
+        if not any([len(network_state.dns_nameservers),
+                    len(network_state.dns_searchdomains)]):
             return None
         header = _make_header(';')
         content_str = str(content)

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Scott Moser(smoser)** wrote on 2019-09-13T13:20:23.623634+00:00_

Based on the content in the description of the bug (the
previously-existing /etc/resolv.conf), cloud-init should *never* be
writing that file when 'netconf' is involved.  It seems rather that
resolv.conf is generated based on contents of
/etc/sysconfig/network/config .

Maybe in the case that there was no dns information, cloud-init should nto
write /etc/sysconfig/network/config .  It is not obvious to me that there
is a clean solution here, but if we take the policy in cloud-init of "do
not ever render empty dns settings", then that means a system that had old
dns settings but got no dns settings on this new instance id will continue
on with the old settings.  DHCP muddies those waters for sure.  Static
networking with explicitly empty dns settings probably is an edge case.

With regard to having to bring up the interface manually:
> I had to bring it up by executing "ifup eth0" from Azure.py code after
> the network config is applied. This way I was able to ssh into the VM.

I think that Ryan identified some incorrect ordering with cloud-init and
wicked that nees to be fixed.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Ryan Harper(raharper)** wrote on 2019-09-13T14:36:24.397425+00:00_

@Nathan,  definitely an improvement, I'll update my merge proposal.

@Scott

For statically configured DNS settings, writting to /etc/resolv.conf is functional but not optimal on SUSE; doing so disables the netconfig automatic management.  As an improvement, cloud-init should look to write these values as NETCONFIG_ variables in /etc/sysconfig/network/config

https://doc.opensuse.org/documentation/leap/reference/html/book.opensuse.reference/cha.network.html#sec.network.manconf.files.resolv

In the case for DHCP, our current policy allows static DNS (we've also called it "global" dns settings, per type:nameserver in v1 config) to apply to any interface that has static ip configuration but no explicit nameserver/search settings.  Interfaces with DHCP do not get the "global" dns settings applied.

We've room for improvement in netplan (v2) where there are explicit DNS overrides for DHCP:

https://github.com/CanonicalLtd/netplan/blob/master/doc/netplan.md#dhcp-overrides

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Moustafa Moustafa(momousta)** wrote on 2019-09-13T17:49:38.268115+00:00_

Actually cloud-init.service has "Before=wicked.service" when I changed it to "After=wicked.service", the eth0 interface was brought up without me explicitly bringing it up. I'm attaching the logs for that VM.
When I checked the "cloud-init.service" file in cloud-init 18.2, It had "After=wicked.service" but in 18.5 and 19.1 it has "Before=wicked.service". I brought up this point before in an offline discussion with Robert and he indicated that in 18.5 a dependency of cloud-init on dhcp-client was added to support the process flow intended by upstream and support dynamic network configuration via cloud-init.
@Robert, can you please elaborate more on that?
_Launchpad attachments_: [cloud-init.tar.gz](https://api.launchpad.net/devel/cloud-init/+bug/1843634/+attachment/5288526/data)

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Ryan Harper(raharper)** wrote on 2019-09-13T18:31:11.153295+00:00_

That's this commit.

commit e0084a599d5259e507298175dad8a1e719017895
Author: Robert Schweikert <rjschwei@suse.com>
Date:   Wed Oct 24 14:34:50 2018 +0000

    systemd: On SUSE ensure cloud-init.service runs before wicked
    
    With db50bc0d9 the sysconfig renderer was enabled for
    openSUSE and SUSE Linux Enterprise. This requires that
    cloud-init.service starts before wicked.
    
    LP: #1799709

diff --git a/systemd/cloud-init.service.tmpl b/systemd/cloud-init.service.tmpl
index b92e8ab..5cb0037 100644
--- a/systemd/cloud-init.service.tmpl
+++ b/systemd/cloud-init.service.tmpl
@@ -14,8 +14,7 @@ After=networking.service
 After=network.service
 {% endif %}
 {% if variant in ["suse"] %}
-Requires=wicked.service
-After=wicked.service
+Before=wicked.service
 # setting hostname via hostnamectl depends on dbus, which otherwise
 # would not be guaranteed at this point.
 After=dbus.service


How cloud-init wants things to work is:

1) cloud-init-local.service runs before networking is up and then will use dhclient (dhcp-client) to temporarily bring up networking to crawl the IMDS to obtain the platform specific network config and write this out to the system

2) os networking starts/comes up

3) cloud-init.service waits for networking to be up runs right after the service, but before network-online.target


Cloud-init invokes dhclient directly in (1) and explicitly does not leave any state from running dhcp around in case that IMDS were to provide a conflicting network configuration; this also avoids starting OS networking services (like wicked) until cloud-init-local has written out the expected network config.

I'll let Robert respond here when he can, but I think we'd want to revert the commit above.  The referenced bug number doesn't have any further details about what was or was not working without it.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Robert Schweikert(rjschwei)** wrote on 2019-09-23T11:42:34.581052+00:00_

I do not remember the details of the change in cloud-init.service. It could be a misunderstanding of the way things were/are supposed to be ordered, then again having "After=" for the other distros is pretty darn obvious, or it could be that a loop is created. Although the creation of a loop between "After=dbus.service" and "After=wicked.service" is hard to imagine, but it will have to be tested.

The creation and wording of lp#1799709 does suggest confusion :(

In any event this is not really related to the problem that we have no DNS information. WHich this bug is really about. cloud-init should not write a resolv.conf file that is "empty" i.e. only has the comments. That's the first problem that needs to be solved.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Robert Schweikert(rjschwei)** wrote on 2019-09-23T12:08:33.549606+00:00_

The patch that is in the 19.1 package that is currently distributed in openSUSE and SLE to avoid writing an "empty" config file is as follows:

--- cloudinit/net/sysconfig.py.orig
+++ cloudinit/net/sysconfig.py
@@ -545,6 +545,8 @@ class Renderer(renderer.Renderer):
             content.add_nameserver(nameserver)
         for searchdomain in network_state.dns_searchdomains:
             content.add_search_domain(searchdomain)
+        if not str(content):
+            return str(content)
         header = _make_header(';')
         content_str = str(content)
         if not content_str.startswith(header):
@@ -658,7 +660,8 @@ class Renderer(renderer.Renderer):
             dns_path = util.target_path(target, self.dns_path)
             resolv_content = self._render_dns(network_state,
                                               existing_dns_path=dns_path)
-            util.write_file(dns_path, resolv_content, file_mode)
+            if resolv_content:
+                util.write_file(dns_path, resolv_content, file_mode)
         if self.networkmanager_conf_path:
             nm_conf_path = util.target_path(target,
                                             self.networkmanager_conf_path)


This is basically the same as in in "master". In "master" instead of returning an empty string, content is initialized to resolv_conf.ResolvConf(""), None is returned. This means in master "_render_dns()" has different return types, which is a style choice and that's a different discussion.

From a code perspective the package delivers the equivalent to what is in master. Next step for me is to do some testing in Azure.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Robert Schweikert(rjschwei)** wrote on 2019-09-23T12:37:50.549949+00:00_

Suitably modified cloudinit/net/sysconfig.py on an instance produces

2019-09-23 12:27:57,994 - sysconfig.py[INFO]: RJS no content for resolv

and renders no resolv.conf file after running 'systemctl start cloud-init-local.service'. Will need access to an instance that shows the behavior per the initial description.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Robert Schweikert(rjschwei)** wrote on 2019-09-23T12:44:15.201096+00:00_

And as a side note comment #13 was incorrect, w.r.t. writing resolv.conf there is no special patch, I was looking at an older package version. Meaning the 19.1 package distributed for openSUSE and SLE  runs the upstream code and returns "None" when there is no dns content in the network config.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Moustafa Moustafa(momousta)** wrote on 2019-09-23T16:51:13.420621+00:00_

The issue is that the content string is reinitialized with the contents of the already existing "resolv.conf" file as following. But the file only contains commented line.

        if existing_dns_path and os.path.isfile(existing_dns_path):
            content = resolv_conf.ResolvConf(util.load_file(existing_dns_path))

So even though there is no name servers nor search domains added to the content, It's still not empty. And the check (not str(content)) will return false. The header will be added to the content and the file will be modified. So netconfig is not touching it.

        if not str(content):
            return None

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Ryan Harper(raharper)** wrote on 2019-09-23T20:30:15.275723+00:00_

summary:  ensuring cloud-init.service runs after wicked and avoid updating resolv.conf without dns values in network_config will fix this issue.


I'm confident that if cloud-init.service runs after wicked.service and the proposed fixed attached to this bug will resolve this issue.  I just wanted to make sure we weren't missing a scenario in reasoning behind wicked running before.

>  or it could be that a loop is created. Although the creation of a loop between
> "After=dbus.service" and "After=wicked.service" is hard to imagine, but it will have
> to be tested.

That could be an issue;  if so, you likely can replace dbus.service with dbus.socket; as this allows the services to initiate connections to dbus.  This is usually sufficient for a dependent service to get going.

As Moustafa pointed out, we need a better patch (which I've proposed and linked to this bug above) which checks if the network config has any dns information before attempting to update resolv.conf;  it appears that in some SLES images they contain a resolv.conf template which is used by 'netconfig' and the cloud-init comment added to the file causes netconfig to not update resolv.conf, meaning no DNS at all.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Robert Schweikert(rjschwei)** wrote on 2019-09-24T19:42:20.509226+00:00_

The attached patch LGTM, I have added it to our current package in OBS [1]. I am currently testing the ordering change.


[1] https://build.opensuse.org/package/show/Cloud:Tools/cloud-init

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Robert Schweikert(rjschwei)** wrote on 2019-09-24T20:11:22.685650+00:00_

Patch tested and order change tested. Both work, found new independent issue in Azure data source.

AFAIK, the attached patch should be landed and we can close this bug.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Robert Schweikert(rjschwei)** wrote on 2019-09-24T20:58:38.644382+00:00_

Order change for wicked: https://code.launchpad.net/~rjschwei/cloud-init/+git/cloud-init/+merge/373166

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Chad Smith(chad.smith)** wrote on 2019-12-19T23:00:32.606977+00:00_

This bug is believed to be fixed in cloud-init in version 19.2-50. If this is still a problem for you, please make a comment and set the state back to New

Thank you.
